### ISSUE TITLE
Which throwM to implement MonadThrow?

### ISSUE THREAD

**Issue by afcowie** [fpco/safe-exceptions#31]
Ideally, you'd be able to
```haskell
 import Control.Exception.Safe
```
and be on your way; "drop-in replacement" or whatever. To that end, **safe-exceptions** re-exports MonadThrow, which is nice. It does not, however, export the `throwM` _method_ of MonadThrow, instead having its own `throwM` function (which is just a convenience alias for `throw`).

The catch is that if you need to implement your own instance of MonadThrow,

```haskell
instance MonadThrow Program where
      throwM = liftIO . throwM
```

you end up with 
```
src/Program.hs:108:5-10: error:
    ‘throwM’ is not a (visible) method of class ‘MonadThrow’
    |
108 |     throwM = liftIO . throwM
    |     ^^^^^^
```

which is a tad inconvenient. I was about to send you a pull request changing the export to `MonadThrow(throwM)` when I realized that there is already a `throwM` and its not the class method. So I'm not sure what to do here. Having to do

```haskell
import Control.Exception.Safe (MonadThrow) hiding (throwM)
import Control.Exception (MonadThrow(throwM))
```
seems likely to not be what you had in mind. How can I help put this right?

AfC

**Comment by snoyberg** [comment]
I don't know of a better solution to the problem. Do you have any
recommendations? I do not want to expose the original throwM, since it
allows throwing async exceptions as sync exceptions.

**Comment by afcowie** [comment]
Ah. I see the problem.

Does it have to be MonadThrow from **exceptions** Control.Monad.Catch? I mean, given that **safe-exceptions** is supposed to be a replacement / best practice / opinionated whatever can we just make a MonadThrow (and I'm assuming MonadCatch etc) that has the same shape but is homed in your library rather than re-exported from the other one?

**Comment by afcowie** [comment]
So what I did for now is:

```haskell
import qualified Control.Exception.Safe as Safe (bracket, throw)
import qualified Control.Monad.Catch as Original (MonadThrow(throwM))

...

instance Original.MonadThrow Program where
    throwM = liftIO . Safe.throw

```
which compiles and appears to run right. I'd still vote for you implementing your own MonadThrow in **safe-exceptions** if that were to be an option.

**Comment by afcowie** [comment]
Hey sure. So close this, obviously, but is what I wrote [above](https://github.com/fpco/safe-exceptions/issues/31#issuecomment-412787343) the right way to work around this, and if so, do you want me to update the documentation?

**Comment by snoyberg** [comment]
Yes, that approach should work, and a doc PR would be welcome. I'll leave this issue open for the doc aspect.
